core/vdbe/mod.rs

```diff
@@ -1011,6 +1028,196 @@ impl Program {
                     }
                     state.pc += 1;
                 }
+                Insn::BitAnd { lhs, rhs, dest } => {
+                    let lhs = *lhs;
+                    let rhs = *rhs;
+                    let dest = *dest;
+                    match (&state.registers[lhs], &state.registers[rhs]) {
+                        // handle 0 and null cases
+                        (OwnedValue::Null, _) | (_, OwnedValue::Null) => {
+                            state.registers[dest] = OwnedValue::Null;
+                        }
+                        (_, OwnedValue::Integer(0))
+                        | (OwnedValue::Integer(0), _)
+                        | (_, OwnedValue::Float(0.0))
+                        | (OwnedValue::Float(0.0), _) => {
+                            state.registers[dest] = OwnedValue::Integer(0);
+                        }
+                        (OwnedValue::Integer(lh), OwnedValue::Integer(rh)) => {
+                            state.registers[dest] = OwnedValue::Integer(lh & rh);
+                        }
+                        (OwnedValue::Float(lh), OwnedValue::Float(rh)) => {
+                            state.registers[dest] = OwnedValue::Integer(*lh as i64 & *rh as i64);
+                        }
+                        (OwnedValue::Float(lh), OwnedValue::Integer(rh)) => {
+                            state.registers[dest] = OwnedValue::Integer(*lh as i64 & rh);
```

Comment:
I'm curious about these operations. I imagined bitwise operations between float and int would not require casting of a number but a transmutation. To explain my point I'll just leave some code:

```rust
fn main() {
    let float_num: f64 = 5.0;
    let int_num: i64 = float_num as i64;

    println!("bits change when casting {}", float_num);
    println!("{:064b}", float_num.to_bits()); // f64 bits
    println!("{:064b}", int_num); // i64 bits

    unsafe {
        let int_num: u64 = std::mem::transmute(float_num);
        println!("bits didn't change when casting {}", float_num);
        println!("{:064b}", float_num.to_bits());
        println!("{:064b}", int_num);
    }
}
```

```bash
bits change when casting 5
0100000000010100000000000000000000000000000000000000000000000000
0000000000000000000000000000000000000000000000000000000000000101
bits didn't change when casting 5
0100000000010100000000000000000000000000000000000000000000000000
0100000000010100000000000000000000000000000000000000000000000000

```

Reply:
This also applies to BitOr and BitNot

Reply:
Looks like sqlite does the same casting so everything good:
```c
/* Convert a floating point value to its closest integer.  Do so in
** a way that avoids 'outside the range of representable values' warnings
** from UBSAN.
*/
i64 sqlite3RealToI64(double r){
  if( r<-9223372036854774784.0 ) return SMALLEST_INT64;
  if( r>+9223372036854774784.0 ) return LARGEST_INT64;
  return (i64)r;
}
```